ddtrace/internal/_encoding.pyx

```diff
@@ -402,3 +606,101 @@ cdef class MsgpackEncoder(MsgpackEncoderBase):
+        self._st = MsgpackStringTable(max_size)
+
+    cpdef flush(self):
+        try:
+            self._st.append_raw(PyLong_FromLong(<long> self.get_buffer()), <Py_ssize_t> super(MsgpackEncoderV05, self).size)
+            return self._st.flush()
+        finally:
+            self._reset_buffer()
+
+    @property
+    def size(self):
+        """Return the size in bytes of the encoder buffer."""
+        return self._st.size + super(MsgpackEncoderV05, self).size
+
+    cpdef put(self, list trace):
+        try:
+            self._st.savepoint()
+            super(MsgpackEncoderV05, self).put(trace)
+        except Exception:
+            self._st.rollback()
+            raise
+
+    cdef inline int _pack_string(self, object string):
+        return msgpack_pack_uint32(&self.pk, self._st._index(string))
+
+    cdef void * get_dd_origin_ref(self, str dd_origin):
+        return <void *> PyLong_AsLong(self._st._index(dd_origin))
+
+    cdef int pack_span(self, object span, void *dd_origin):
+        cdef int ret
+
+        ret = msgpack_pack_array(&self.pk, 12)
+        if ret != 0: return ret
+
+        ret = self._pack_string(span.service)
+        if ret != 0: return ret
+        ret = self._pack_string(span.name)
+        if ret != 0: return ret
+        ret = self._pack_string(span.resource)
+        if ret != 0: return ret
+
+        _ = span.trace_id
+        ret = msgpack_pack_uint64(&self.pk, _ if _ is not None else 0)
+        if ret != 0: return ret
+        
+        _ = span.span_id
+        ret = msgpack_pack_uint64(&self.pk, _ if _ is not None else 0)
+        if ret != 0: return ret
+        
+        _ = span.parent_id
+        ret = msgpack_pack_uint64(&self.pk, _ if _ is not None else 0)
+        if ret != 0: return ret
+        
+        _ = span.start_ns
+        ret = msgpack_pack_int64(&self.pk, _ if _ is not None else 0)
+        if ret != 0: return ret
+        
+        _ = span.duration_ns
+        ret = msgpack_pack_int64(&self.pk, _ if _ is not None else 0)
+        if ret != 0: return ret
+        
+        _ = span.error
+        ret = msgpack_pack_int32(&self.pk, _ if _ is not None else 0)
+        if ret != 0: return ret
+
+        ret = msgpack_pack_map(&self.pk, len(span.meta) + (dd_origin is not NULL))
+        if ret != 0: return ret
+        if span.meta:
+            for k, v in span.meta.items():
+                ret = self._pack_string(k)
+                if ret != 0: return ret
+                ret = self._pack_string(v)
+                if ret != 0: return ret
+        if dd_origin is not NULL:
+            ret = msgpack_pack_uint32(&self.pk, <stdint.uint32_t> 1)
+            if ret != 0: return ret
+            ret = msgpack_pack_uint32(&self.pk, <stdint.uint32_t> dd_origin)
+            if ret != 0: return ret
+        
+        ret = msgpack_pack_map(&self.pk, len(span.metrics))
+        if ret != 0: return ret
+        if span.metrics:
+            for k, v in span.metrics.items():
```

Comment:
random aside: I wonder if adding `cdef str metrics_key; cdef long metrics_value;` up above would help here? doubt it, since the same number of type checking/conversions would probably happen, it would just get moved to the iteration step rather than the `_pack_string/_pack_number` calls.